csrc/python_frontend/translation.cpp

```diff
@@ -0,0 +1,524 @@
+        std::back_inserter(shape),
+        [](IterDomain* id) {
+          return (id->getMaybeExpandedExtent()->isConstScalar())
+              ? id->getMaybeExpandedExtent()->evaluate().as<int64_t>()
+              : -1;
+        });
+
+    fd_->defineRecord(new TensorRecord(
+        {fd_->recordingState(output())},
+        shape,
+        tv->domain()->contiguity(),
+        std::get<PrimDataType>(tv->dtype().type),
+        tv->isCpuScalar(),
+        tv->domain()->strideOrder()));
+  }
+
+  // =================================================================================
+  // Handle add_output variants
+
+  // Add Tensor output to FusionDefinition
+  void handleOutput(const TensorView* tv) {
+    size_t output_index = map_val_to_fd_index_.at(tv);
+    fd_->defineRecord(new OutputRecord<TensorView>(
+        {fd_->recordingState(output_index)},
+        serde::RecordType::OutputTv,
+        tv->domain()->strideOrder()));
+  }
+
+  // Alias output Tensor with input tensor
+  void handleOutput(const TensorView* tv, const AliasInfo& alias_info) {
+    size_t output_index = map_val_to_fd_index_.at(tv);
+    size_t input_index = map_val_to_fd_index_.at(alias_info.aliased_io);
+    fd_->defineRecord(new OutputRecord<TensorView>(
+        {fd_->recordingState(output_index), fd_->recordingState(input_index)},
+        serde::RecordType::OutputTv));
+  }
+
+  // =================================================================================
+  // Map CPP Expression classes to corresponding RecordFunctors in
+  // python_frontend
+
+  // A generic function to map UnaryOp, BinaryOp, and TernaryOp to
+  // python_frontend OpRecord
+  template <typename ExprType, typename ResultType, typename... ArgTypes>
+  void handleOpRecord(
+      const Expr* e,
+      serde::RecordType record_type,
+      ResultType result,
+      ArgTypes... args) {
+    NVF_ERROR(e->isA<ExprType>());
+    std::vector<State> argument_states;
+    std::transform(
+        e->inputs().begin(),
+        e->inputs().end(),
+        std::back_inserter(argument_states),
+        [&](auto arg) {
+          return fd_->recordingState(map_val_to_fd_index_.at(arg));
+        });
+
+    fd_->defineRecord(new OpRecord<ResultType, ArgTypes...>(
+        argument_states,
+        {fd_->recordingState(map_val_to_fd_index_.at(result))},
+        "ops." + getString(e->as<ExprType>()),
+        record_type,
+        getFunction<ResultType, ArgTypes...>(e->as<ExprType>())));
+  }
+
+  // Map UnaryOp to python_frontend OpRecord
+  void handle(const UnaryOp* uop) final {
+    // TODO Add support for other unary ops
+    // Handle only Cast operation
+    NVF_ERROR(uop->getUnaryOpType() == UnaryOpType::Cast);
+    handleCastOp(uop);
+  }
+
+  // Map cast UnaryOp to CastOpRecord
+  void handleCastOp(const Expr* op) {
+    bool is_cast_op = op->isA<UnaryOp>() &&
+        op->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Cast;
+    NVF_ERROR(op->isA<LoadStoreOp>() || is_cast_op);
```

Comment:
Why is `LoadStoreOp` accepted here?

Reply:
If the dtype between input and output are the same, `castOp` creates a `LoadStoreOp`. See https://github.com/NVIDIA/Fuser/blob/main/csrc/ops/arith.cpp#L30-L34.

If we see a `LoadStoreOp` without permutation, we map back to a `castOp`.

Reply:
`Broadcast` also creates a `set` if it is a no-op.

Reply:
I see. So we never map back to `fd.ops.set` then. I wonder if we should just skip translating trivial sets unless they are `segment_set`.